optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,634 @@
+            # So we can ignore the case.
+
+            # The cache key is calculated based on the key of the previous generation and
+            # the remaining running trials in the current population.
+            # If there are no running trials, the new cache key becomes exactly the same as
+            # the previous one, and the cached content will be overwritten. This allows us to
+            # skip redundant cache key calculations when this method is called for the subsequent
+            # trials.
+            for trial in generation_to_runnings[generation]:
+                hasher.update(bytes(str(trial.number), "utf-8"))
+
+            cache_key = "{}:{}".format(_POPULATION_CACHE_KEY_PREFIX, hasher.hexdigest())
+            study_system_attrs = study._storage.get_study_system_attrs(study._study_id)
+            cached_generation, cached_population_numbers = study_system_attrs.get(
+                cache_key, (-1, [])
+            )
+            if cached_generation >= generation:
+                generation = cached_generation
+                population = [trials[n] for n in cached_population_numbers]
+            else:
+                population.extend(parent_population)
+                population = self._select_elite_population(study, population)
+
+                # To reduce the number of system attribute entries,
+                # we cache the population information only if there are no running trials
+                # (i.e., the information of the population has been fixed).
+                # Usually, if there are no too delayed running trials, the single entry
+                # will be used.
+                if len(generation_to_runnings[generation]) == 0:
+                    population_numbers = [t.number for t in population]
+                    study._storage.set_study_system_attr(
+                        study._study_id, cache_key, (generation, population_numbers)
+                    )
+
+            parent_generation = generation
+            parent_population = population
+
+        return parent_generation, parent_population
+
+    def _select_elite_population(
+        self, study: Study, population: list[FrozenTrial]
+    ) -> list[FrozenTrial]:
+        elite_population: list[FrozenTrial] = []
+        population_per_rank = self._fast_non_dominated_sort(population, study.directions)
+        for population in population_per_rank:
+            if len(elite_population) + len(population) < self._population_size:
+                elite_population.extend(population)
+            else:
+                n_objectives = len(study.directions)
+                # Construct reference points in the first run.
+                if self.reference_points is None:
+                    self.reference_points = _generate_default_reference_point(
+                        n_objectives, self._dividing_parameter
+                    )
```

Comment:
It would be great to construct the reference point in the `__init__`.

Reply:
IIRC, we came to the conclusion that `n_objectives` is necessary to construct default reference points and it is generally unknown when constructing sampler.

Reply:
Thanks for the reply. I understood the reason. Let me resolve this comment.